Account for static members in secret replacement

__What should we do?__

Account for static members in classes when replacing secrets.

Our current take on this:

* You can have static fields in your class.
* You can have static methods in your class.
* We allow retrieving values via static methods.
* We do not allow replacing secrets on static fields.

__Why should we do it?__

Java classes can contain dynamically (but also manually by the user) created `static` members (fields or methods). Those cannot be probed using `canAccess` with an actual non-null input object. They expect to be probed with a `null` object.

_Originally posted by @igpetrov in https://github.com/camunda/connector-sdk/issues/69#issuecomment-1256443283_

Hi @jonathanlukas @tmetzke,

While investigating a failing test at https://github.com/camunda/connector-slack/pull/90, I found a potential bug and would like to validate with both of you.

Given [code snippet in the Slack connector](https://github.com/camunda/connector-slack/blob/main/src/main/java/io/camunda/connector/slack/ChatPostMessageData.java#L71-L76), it generates anonymous lambda function under the hood. See the last method on the screenshot:

<img width="1155" alt="Screenshot 2022-09-23 at 19 33 58" src="https://user-images.githubusercontent.com/108870003/192010981-8067194d-343a-4673-9db2-d2e341d2ede5.png">

This leads to an exception `java.lang.RuntimeException: java.lang.IllegalArgumentException: non-null object for private static boolean io.camunda.connector.slack.ChatPostMessageData.lambda$getUserId$0(java.lang.String,com.slack.api.model.User)`:

<details><summary>Click to expand</summary>

```
java.lang.RuntimeException: java.lang.IllegalArgumentException: non-null object for private static boolean io.camunda.connector.slack.ChatPostMessageData.lambda$getUserId$0(java.lang.String,com.slack.api.model.User)

	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.setProperty(AbstractOutboundConnectorContext.java:234)
	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.lambda$handleSecretsField$4(AbstractOutboundConnectorContext.java:115)
	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.lambda$handleSecretsElement$6(AbstractOutboundConnectorContext.java:127)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.handleSecretsElement(AbstractOutboundConnectorContext.java:121)
	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.lambda$handleSecretsField$5(AbstractOutboundConnectorContext.java:107)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.handleSecretsField(AbstractOutboundConnectorContext.java:105)
	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.replaceSecrets(AbstractOutboundConnectorContext.java:58)
	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.lambda$handleSecretsElement$6(AbstractOutboundConnectorContext.java:124)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.handleSecretsElement(AbstractOutboundConnectorContext.java:121)
	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.lambda$handleSecretsField$5(AbstractOutboundConnectorContext.java:107)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.handleSecretsField(AbstractOutboundConnectorContext.java:105)
	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.replaceSecrets(AbstractOutboundConnectorContext.java:58)
	at io.camunda.connector.slack.SlackConnectorRequestTest.replaceSecrets_shouldReplaceSecrets(SlackConnectorRequestTest.java:77)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:217)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:57)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: java.lang.IllegalArgumentException: non-null object for private static boolean io.camunda.connector.slack.ChatPostMessageData.lambda$getUserId$0(java.lang.String,com.slack.api.model.User)
	at java.base/java.lang.reflect.AccessibleObject.canAccess(AccessibleObject.java:454)
	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.lambda$setProperty$13(AbstractOutboundConnectorContext.java:211)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:178)
	at java.base/java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:1002)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
	at io.camunda.connector.impl.outbound.AbstractOutboundConnectorContext.setProperty(AbstractOutboundConnectorContext.java:224)
	... 102 more
```

</details>

I think, this is a bug. While we can rewrite [the original code snippet](https://github.com/camunda/connector-slack/blob/main/src/main/java/io/camunda/connector/slack/ChatPostMessageData.java#L71-L76) not to use anonymous lambdas, it doesn't solve the issue when other customers may want to use streams or other functional utilities.

Another option is to introduce additional filter statement `.filter(method -> !Modifier.isStatic(method.getModifiers()))` [here](https://github.com/camunda/connector-sdk/blob/main/core/src/main/java/io/camunda/connector/impl/outbound/AbstractOutboundConnectorContext.java#L208) but I am not sure whether we would face with any side-effects.

WDYT?

cc: @Oleksiivanov

--- a/core/src/main/java/io/camunda/connector/impl/outbound/AbstractOutboundConnectorContext.java
+++ b/core/src/main/java/io/camunda/connector/impl/outbound/AbstractOutboundConnectorContext.java
@@ -145,9 +145,10 @@ public abstract class AbstractOutboundConnectorContext implements OutboundConnec
 
   @SuppressWarnings("unchecked")
   protected static <T> T getProperty(Object input, Field field) {
-    if (field.canAccess(input)) {
+    final var inputToCall = Modifier.isStatic(field.getModifiers()) ? null : input;
+    if (field.canAccess(inputToCall)) {
       try {
-        return (T) field.get(input);
+        return (T) field.get(inputToCall);
       } catch (IllegalAccessException e) {
         throw new RuntimeException(e);
       }
@@ -156,7 +157,9 @@ public abstract class AbstractOutboundConnectorContext implements OutboundConnec
       Method getter =
           Arrays.stream(field.getDeclaringClass().getDeclaredMethods())
               // method has to be public
-              .filter(method -> method.canAccess(input))
+              .filter(
+                  method ->
+                      method.canAccess(Modifier.isStatic(method.getModifiers()) ? null : input))
               // method has to follow java getter conventions
               .filter(
                   method ->
@@ -196,6 +199,13 @@ public abstract class AbstractOutboundConnectorContext implements OutboundConnec
               + "' of type "
               + field.getDeclaringClass());
     }
+    if (Modifier.isStatic(field.getModifiers())) {
+      throw new IllegalStateException(
+          "Cannot invoke set or setter on static field '"
+              + field.getName()
+              + "' of type "
+              + field.getDeclaringClass());
+    }
     if (field.canAccess(input)) {
       try {
         field.set(input, property);
@@ -206,6 +216,8 @@ public abstract class AbstractOutboundConnectorContext implements OutboundConnec
     try {
       Method setter =
           Arrays.stream(field.getDeclaringClass().getDeclaredMethods())
+              // method must not be static
+              .filter(method -> !Modifier.isStatic(method.getModifiers()))
               // method has to be public
               .filter(method -> method.canAccess(input))
               // method has to follow java setter conventions

--- a/core/src/test/java/io/camunda/connector/impl/outbound/OutboundConnectorContextTest.java
+++ b/core/src/test/java/io/camunda/connector/impl/outbound/OutboundConnectorContextTest.java
@@ -52,7 +52,10 @@ class OutboundConnectorContextTest {
               new GetPropertyTestInput("protectedField", testInput.protectedField),
               new GetPropertyTestInput("privateField", testInput.getPrivateField()),
               new GetPropertyTestInput("finalField", testInput.finalField),
-              new GetPropertyTestInput("packageField", testInput.packageField));
+              new GetPropertyTestInput("packageField", testInput.packageField),
+              new GetPropertyTestInput("staticField", OutboundTestInput.staticField),
+              new GetPropertyTestInput(
+                  "privateStaticField", OutboundTestInput.getPrivateStaticField()));
       ThrowingConsumer<GetPropertyTestInput> executor =
           in -> {
             // when
@@ -102,6 +105,44 @@ class OutboundConnectorContextTest {
           .isInstanceOf(IllegalStateException.class)
           .hasMessage(
               "Cannot invoke set or setter on final field 'finalField' of type class "
+                  + testInput.getClass().getName());
+    }
+
+    @Test
+    public void shouldThrowWhenSettingOnStaticField() {
+      // given
+      OutboundTestInput testInput = new OutboundTestInput();
+      String modified = "modified";
+      Exception expected =
+          catchException(
+              () ->
+                  // when
+                  AbstractOutboundConnectorContext.setProperty(
+                      testInput, fieldForName("staticField"), modified));
+      // then
+      assertThat(expected)
+          .isInstanceOf(IllegalStateException.class)
+          .hasMessage(
+              "Cannot invoke set or setter on static field 'staticField' of type class "
+                  + testInput.getClass().getName());
+    }
+
+    @Test
+    public void shouldThrowWhenSettingOnPrivateStaticField() {
+      // given
+      OutboundTestInput testInput = new OutboundTestInput();
+      String modified = "modified";
+      Exception expected =
+          catchException(
+              () ->
+                  // when
+                  AbstractOutboundConnectorContext.setProperty(
+                      testInput, fieldForName("privateStaticField"), modified));
+      // then
+      assertThat(expected)
+          .isInstanceOf(IllegalStateException.class)
+          .hasMessage(
+              "Cannot invoke set or setter on static field 'privateStaticField' of type class "
                   + testInput.getClass().getName());
     }
   }
@@ -291,7 +332,7 @@ class OutboundConnectorContextTest {
     }
 
     @Test
-    void shouldReplaceSecretsInImmutableStringMap() {
+    void shouldFailIfReplaceSecretsInImmutableStringMap() {
       // given
       OutboundConnectorContext connectorContext =
           OutboundConnectorContextBuilder.create().secret("s3cr3t", "plain").build();
@@ -317,6 +358,34 @@ class OutboundConnectorContextTest {
           .isInstanceOf(IllegalStateException.class)
           .hasMessageContaining("Element at key")
           .hasMessageContaining("in map has no nested properties and is no String!");
+    }
+
+    @Test
+    void shouldFailIfReplaceSecretsInStaticString() {
+      // given
+      OutboundConnectorContext connectorContext =
+          OutboundConnectorContextBuilder.create().secret("s3cr3t", "plain").build();
+      final var testInput = new InputStaticString();
+      // when
+      Exception expected = catchException(() -> connectorContext.replaceSecrets(testInput));
+      // then
+      assertThat(expected)
+          .isInstanceOf(IllegalStateException.class)
+          .hasMessageContaining("Cannot invoke set or setter on static field");
+    }
+
+    @Test
+    void shouldFailIfReplaceSecretsInPrivateStaticString() {
+      // given
+      OutboundConnectorContext connectorContext =
+          OutboundConnectorContextBuilder.create().secret("s3cr3t", "plain").build();
+      final var testInput = new InputPrivateStaticString();
+      // when
+      Exception expected = catchException(() -> connectorContext.replaceSecrets(testInput));
+      // then
+      assertThat(expected)
+          .isInstanceOf(IllegalStateException.class)
+          .hasMessageContaining("Cannot invoke set or setter on static field");
     }
 
     @Test
@@ -467,5 +536,17 @@ class OutboundConnectorContextTest {
 
   static class InputStringSet {
     @Secret public Set<String> stringSet = new HashSet<>(Set.of("secrets.s3cr3t", "foo"));
+  }
+
+  static class InputStaticString {
+    @Secret public static String staticString = "secrets.s3cr3t";
+  }
+
+  static class InputPrivateStaticString {
+    @Secret private static String staticString = "secrets.s3cr3t";
+
+    public static String getStaticString() {
+      return staticString;
+    }
   }
 }

--- a/core/src/test/java/io/camunda/connector/impl/outbound/OutboundTestInput.java
+++ b/core/src/test/java/io/camunda/connector/impl/outbound/OutboundTestInput.java
@@ -19,11 +19,14 @@ package io.camunda.connector.impl.outbound;
 import io.camunda.connector.api.annotation.Secret;
 
 public class OutboundTestInput {
+
+  public static String staticField = "static";
   public final String finalField = "final";
   public String publicField = "public";
   protected String protectedField = "protected";
   String packageField = "package";
   private String privateField = "private";
+  private static String privateStaticField = "privateStatic";
 
   @Secret private String secretField = "secrets.s3cr3t";
 
@@ -69,5 +72,13 @@ public class OutboundTestInput {
 
   public void setSecretField(String secretField) {
     this.secretField = secretField;
+  }
+
+  public static void setPrivateStaticField(String staticFieldValue) {
+    privateStaticField = staticFieldValue;
+  }
+
+  public static String getPrivateStaticField() {
+    return privateStaticField;
   }
 }

